src/java/org/apache/cassandra/io/util/TrackedDataInputPlus.java

```diff
@@ -129,22 +156,34 @@ public String readUTF() throws IOException
 
     public int readUnsignedByte() throws IOException
     {
+        checkCanRead(1);
         int i = source.readUnsignedByte();
         bytesRead += 1;
         return i;
     }
 
     public int readUnsignedShort() throws IOException
     {
+        checkCanRead(2);
         int i = source.readUnsignedShort();
         bytesRead += 2;
         return i;
     }
 
     public int skipBytes(int n) throws IOException
     {
-        int skipped = source.skipBytes(n);
+        int skipped = source.skipBytes(limit < 0 ? n : (int) Math.min(limit - bytesRead, n));
         bytesRead += skipped;
         return skipped;
     }
+
+    @Inline
+    private void checkCanRead(int size) throws IOException
+    {
+        if (limit >= 0 && bytesRead + size > limit)
```

Comment:
will limit just be zero？

Reply:
I just test some times,it seems limit can not be 0 ~~~

Reply:
Yes, we always encode at least a flag